### PR TITLE
Use `_prime_post_caches` if the query is not filtered.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/58599

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-query.php [hunk 1]
```diff
@@ -3270,10 +3270,11 @@ public function get_posts() {
 			return $post_parents;
 		}
 
+		$unfiltered_query = $old_request == $this->request && "{$wpdb->posts}.*" === $fields;
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php]: Two things:
* Why a non-strict comparison here? Can't we use `===`?
* Nit-pick, but maybe `$is_unfiltered_query` is a better name here, to clarify that this is a boolean.

File: src/wp-includes/class-wp-query.php [hunk 2]
```diff
@@ -3529,7 +3530,12 @@ public function get_posts() {
 			$this->posts = array_map( 'get_post', $this->posts );
 
 			if ( $q['cache_results'] ) {
-				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+				if ( $unfiltered_query ) {
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php]: Per my original feedback in https://github.com/WordPress/wordpress-develop/pull/5134#discussion_r1330566960: We also have to ensure the query results were not modified, by storing the value from https://github.com/WordPress/wordpress-develop/blob/3bacfc9f978876549ee0e2b5068be02e006d956c/src/wp-includes/class-wp-query.php#L3337 in a variable, e.g. `$unfiltered_posts`.

Then we can check it here like this:
```suggestion
				if ( $unfiltered_query && $this->posts === $unfiltered_posts ) {
```

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey Only one crucial point of feedback here, similar to what I mentioned on the original PR.

Other than that, this looks good to me.

File: src/wp-includes/class-wp-query.php (hunk 2 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php]: I see now. Good catch. Made the tweak https://github.com/WordPress/wordpress-develop/pull/5251/commits/f575f45ae1505271eaa59d5753f10a7269917c4a.

File: src/wp-includes/class-wp-query.php (hunk 1 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php]: I have changed the variable name. But I will not change the logic for now. That is the logic it was before and I don't want to change it.

Comment (felixarntz) [reply src/wp-includes/class-wp-query.php]: @spacedmonkey Why not use `===`? I see this was using `==` before, but I don't think there's a good reason not to fix it. The request is always a string, and I don't think for strings the `==` or `===` even makes a real difference. I think we should use `===` in accordance with the latest WordPress coding standards and using strict comparisons.

**Review by felixarntz — APPROVED** [review]
Thanks @spacedmonkey for the updates, LGTM. I'd still recommend using a strict comparison in https://github.com/WordPress/wordpress-develop/pull/5251#discussion_r1331781305, but not a blocker.

Would be great to get @peterwilsoncc's review on this as well before committing.

**Comment by spacedmonkey** [comment]
Committed https://github.com/WordPress/wordpress-develop/commit/e486ac07458d36182ee75091eac06f9a99a5ae2a

**Pull request closed without merging**